`Development`: Add application table for applicants - client side

Douglas-Sarah opened the pull request:
<!-- Thanks for contributing to TumApply! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist

#### General

<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->

- [x] I tested **all** changes and their related features with **all** corresponding user types.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://confluence.aet.cit.tum.de/spaces/AP/pages/257786006/Language+Guidelines+Client).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311714/PR+Guidelines).

#### Server

- [x] **Important**: I implemented the changes with a [very good performance](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311718/Performance+Guidelines) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311877/Server+Guidelines).
- [x] I documented the Java code using JavaDoc style.

#### Client

- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311716/Client+Guidelines).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311925/Client+Test+Guidelines).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
With this PR, the applicant gets a clear and responsive overview of all their job applications so they can easily track their status. This feature allows them to view, update, withdraw, or delete applications efficiently, improving the overall user experience.

### Description

Applicant Application Overview:
- Fetches application data and total count via backend API
- Supports delete and withdraw actions with confirmation
- Automatically reloads the current page after deletion/withdrawal
- When #PR167 is merged, will support editing applications when clicking on 'update' button
- Pagination
- Badge display for application state

Caveats:
- the user is still hardcoded
- view is still disabled, as there is not yet an application-detail-page to route to

### Steps for Testing

<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
0. update database
1. Login as admin1 for example
2. on homepage, click on button "View my Applications"
3. Test whether the table is displayed with 4 applications, with 4 different status's
4. Test whether after a withdraw action, the status is changed to 'withdrawn'
5. Test whether after a delete action, the application is deleted and is not displayed anymore in the table

### Review Progress

<!-- Each PR should be reviewed by at least one other developers. The code, the functionality (= manual test). -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review

- [ ] Code Review 1

#### Manual Tests

- [ ] Test 1

### Screenshots

![image](https://github.com/user-attachments/assets/d11c989e-47a7-4060-8650-ee154ada1598)

<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

MoSchmidt commented on src/main/java/de/tum/cit/aet/application/web/ApplicationResource.java:
```diff
@@ -104,4 +115,18 @@ public ResponseEntity<Void> withdrawApplication(@PathVariable UUID applicationId
+    }
+
+    @GetMapping("/pages/length")
+    public ResponseEntity<Long> getApplicationPagesLength(@RequestParam UUID applicantId) {
```
Is it necessary to send the applicantId as a request parameter? I assume only applicants themself will acces this endpoint? Then this could be read from the SecurityContextHolder. Otherwise we would first retreive the applicantId and also store it in the client

Douglas-Sarah replied:
I agree that it is not necessary here. There is a separate branch where I am working on integrating the Authorization into the Application Module, both client-wise and server-wise, because until now I hardcoded everything.
But for not having to change the signature later on, I will change it and hardcode the applicant Id here as well

marc-fett replied:
But what if an admin accesses this page? Then we would need to add the applicantId since the admin is not the "creator" of this application, right?

MoSchmidt commented on src/main/java/de/tum/cit/aet/application/service/ApplicationService.java:
```diff
@@ -204,6 +207,17 @@ public void withdrawApplication(UUID applicationId) {
     }
+
+    public List<ApplicationOverviewDTO> getAllApplications(UUID applicantId, int pageSize, int pageNumber) {
+        return applicationRepository.findApplicationsByApplicant(applicantId, pageNumber, pageSize);
```
This method is not defined in the `ApplicationRespository`, thus the build fails

Douglas-Sarah replied:
Thanks :) Like I wrote below, I fixed it with a recent commit

MoSchmidt reviewed (changes requested):
The Repository method is not defined inside the `ApplicationRepository` but in `ApplicationEntityRepository`and the server is not compiling. To fix this you just need to extend the `ApplicationEntityRepository` in the normal `ApplicationRepository`. This might also be the problem why so many workflows are failing right now.

I also left one question about the endpoint and why we are sending the applicantId. But if there is a specific reason to do it like that, then it is no problem.

I see that the view button is always disabled. Is this on purpose and will this be changed or is there a reason to it?

Except for those two points the code looks really good and I also appreciate that you added a lot of client tests

ani-ib commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.html:
```diff
@@ -0,0 +1,58 @@
+
+  <ng-template #createdTemplate let-row>
+    <span class="w-full">
+      {{ calculateDate(row.createdAt) }}
```
please use signals to calculate the Date and not method calls in the html

Douglas-Sarah replied:
I moved it to the server. I used your method, maybe it would be useful to have a shared service for that, but that is also something to discuss on monday.

ani-ib commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+  }
+
+  calculateDate(createdDateString: string): string {
```
please use a signal to calculate the date
It would be even better to calculate the date in the server directly if possible; I have done something similar in the JobCardDTO so you can have a look at that.

ani-ib reviewed (changes requested):
Table looks good, thanks for the implementation :) Left some minor comments that need to be adjusted. 
I personally didn't imagine the "Status" tags to be this color scheme, because they look a lot like buttons now which might be confusing to users, but that's something we can all discuss on Monday, so its fine for now

ani-ib reviewed (changes requested):
I found another issue while testing this PR:
The delete functionality does not work for me.
When click on "Delete", I get the following alert which says "error withdrawing application" and not "error deleting application".
Furthermore, the application is not removed from the Database or from the Applications page for me.
I get the following error in console:
<img width="595" alt="image" src="https://github.com/user-attachments/assets/ae9e7765-5fbc-4a3e-b5d4-a1e29d9dd2d8" />
And this is shown in my IDE:
"2025-06-01T12:47:55.731+02:00  WARN 37223 --- [nio-8080-exec-8] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.dao.InvalidDataAccessResourceUsageException: JDBC exception executing SQL [select ic1_0.application_id,ic1_0.internal_comment_id,ic1_0.created_at,cb1_0.user_id,case when cb1_1.user_id is not null then 1 when cb1_0.user_id is not null then 0 end,cb1_0.avatar,cb1_0.birthday,cb1_0.created_at,cb1_0.email,cb1_0.first_name,cb1_0.gender,cb1_0.last_modified_at,cb1_0.last_name,cb1_0.linkedin_url,cb1_0.nationality,cb1_0.phone_number,rg1_0.research_group_id,rg1_0.abbreviation,rg1_0.city,rg1_0.created_at,rg1_0.default_field_of_studies,rg1_0.description,rg1_0.email,rg1_0.head,rg1_0.last_modified_at,rg1_0.name,rg1_0.postal_code,rg1_0.school,rg1_0.street,rg1_0.website,cb1_0.selected_language,cb1_0.website,cb1_1.bachelor_degree_name,cb1_1.bachelor_grade,cb1_1.bachelor_grading_scale,cb1_1.bachelor_university,cb1_1.city,cb1_1.country,cb1_1.master_degree_name,cb1_1.master_grade,cb1_1.master_grading_scale,cb1_1.master_university,cb1_1.postal_code,cb1_1.street,ic1_0.last_modified_at,ic1_0.message from internal_comments ic1_0 left join users cb1_0 on cb1_0.user_id=ic1_0.created_by left join applicants cb1_1 on cb1_0.user_id=cb1_1.user_id left join research_groups rg1_0 on rg1_0.research_group_id=cb1_0.research_group_id where ic1_0.application_id=?] [Table 'tumapply.internal_comments' doesn't exist] [n/a]; SQL [n/a]] 
"

Would it be realistic to get this fixed and merged before Monday, or are you planning on fixing this in another PR?

ani-ib commented:
I have two other comments regarding the design of the table, but they can be discussed on Monday if needed and added to another PR:
1. I think the table would look more coherent if the status tags had the same width in the table
<img width="155" alt="image" src="https://github.com/user-attachments/assets/c2716a57-edea-4b6f-b65c-b70b864cd5ef" />

2. I believe it would also look better if the button widths for each column were the same (e.g. "View" Buttons Column, "Update/Widthdraw" buttons column
<img width="284" alt="image" src="https://github.com/user-attachments/assets/5f0794aa-60c5-40ca-bf17-99cfa121372a" />

Because right now it also does not look coherent in my opinion since the "Update" and "Withdraw" buttons have different sizes

Douglas-Sarah commented:
> I found another issue while testing this PR: The delete functionality does not work for me. When click on "Delete", I get the following alert which says "error withdrawing application" and not "error deleting application". Furthermore, the application is not removed from the Database or from the Applications page for me. I get the following error in console: <img alt="image" width="595" src="https://private-user-images.githubusercontent.com/88331280/449744502-ae9e7765-5fbc-4a3e-b5d4-a1e29d9dd2d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg3NzYxODAsIm5iZiI6MTc0ODc3NTg4MCwicGF0aCI6Ii84ODMzMTI4MC80NDk3NDQ1MDItYWU5ZTc3NjUtNWZiYy00YTNlLWI1ZDQtYTFlMjlkOWRkMmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjAxVDExMDQ0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5Nzc2YThmYzNkYzgxNzMxOTE2MDRkZjE1YzlhMTA1MGJlNDExNjc1OWZjZDY1MjUxOWVjODYzM2Y3MzIyZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3z5Tqtz_-YAdyUw5PqI0rJ3LpCrYTh8KCed7YH2tp0E"> And this is shown in my IDE: "2025-06-01T12:47:55.731+02:00 WARN 37223 --- [nio-8080-exec-8] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.dao.InvalidDataAccessResourceUsageException: JDBC exception executing SQL [select ic1_0.application_id,ic1_0.internal_comment_id,ic1_0.created_at,cb1_0.user_id,case when cb1_1.user_id is not null then 1 when cb1_0.user_id is not null then 0 end,cb1_0.avatar,cb1_0.birthday,cb1_0.created_at,cb1_0.email,cb1_0.first_name,cb1_0.gender,cb1_0.last_modified_at,cb1_0.last_name,cb1_0.linkedin_url,cb1_0.nationality,cb1_0.phone_number,rg1_0.research_group_id,rg1_0.abbreviation,rg1_0.city,rg1_0.created_at,rg1_0.default_field_of_studies,rg1_0.description,rg1_0.email,rg1_0.head,rg1_0.last_modified_at,rg1_0.name,rg1_0.postal_code,rg1_0.school,rg1_0.street,rg1_0.website,cb1_0.selected_language,cb1_0.website,cb1_1.bachelor_degree_name,cb1_1.bachelor_grade,cb1_1.bachelor_grading_scale,cb1_1.bachelor_university,cb1_1.city,cb1_1.country,cb1_1.master_degree_name,cb1_1.master_grade,cb1_1.master_grading_scale,cb1_1.master_university,cb1_1.postal_code,cb1_1.street,ic1_0.last_modified_at,ic1_0.message from internal_comments ic1_0 left join users cb1_0 on cb1_0.user_id=ic1_0.created_by left join applicants cb1_1 on cb1_0.user_id=cb1_1.user_id left join research_groups rg1_0 on rg1_0.research_group_id=cb1_0.research_group_id where ic1_0.application_id=?] [Table 'tumapply.internal_comments' doesn't exist] [n/a]; SQL [n/a]] "
> 
> Would it be realistic to get this fixed and merged before Monday, or are you planning on fixing this in another PR?

Thank you for your review- I already fixed this issue on another branch of mine, so that was probably why the error did not occur for me when testing. I have added the fix here as well. It should work now

ani-ib reviewed (approved):
Code LGTM - table looks well-designed and everything works as intended, thanks for the changes :)

marc-fett reviewed (approved):
Code LGTM. The CSS seems a bit of since the table is centered and the padding is not adapted to the screen size. Meaning when the screen is smaller, the padding is still there. But this is not a show stopper for me 👍🏼

az108 commented on src/main/webapp/app/home/home.component.html:
```diff
@@ -17,6 +17,12 @@ <h1 class="display-4"><span jhiTranslate="home.title">Welcome to TUMApply!</span
       (click)="this.router.navigate(['/application/edit/bce85235-25b7-46c2-988c-9290be79dc57'])"
     />
+    <jhi-button
+      (click)="this.router.navigate(['/application/overview'])"
```
can we extract this to a method in the ts?

az108 commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.spec.ts:
```diff
@@ -0,0 +1,183 @@
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+
+    expect(spy).toHaveBeenCalledWith();
```
toHaveBeenCalledWith() , did you forget something here ?

az108 commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.spec.ts:
```diff
@@ -0,0 +1,183 @@
+    component.onWithdrawApplication('app-id');
+
+    expect(applicationService.withdrawApplication).toHaveBeenCalledWith('app-id');
+    expect(reloadSpy).toHaveBeenCalled();
```
can we be more concrete how often called?

Douglas-Sarah replied:
Solved 👍

az108 commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.spec.ts:
```diff
@@ -0,0 +1,183 @@
+    component.onDeleteApplication('app-id');
+
+    expect(applicationService.deleteApplication).toHaveBeenCalledWith('app-id');
+    expect(reloadSpy).toHaveBeenCalled();
```
same here

Douglas-Sarah replied:
Solved 👍

az108 commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.ts:
```diff
@@ -0,0 +1,109 @@
+  templateUrl: './application-overview-for-applicant.component.html',
+  styleUrl: './application-overview-for-applicant.component.scss',
+})
+export default class ApplicationOverviewForApplicantComponent {
```
we should comment this component a bit more, as not everything here is trivial

Douglas-Sarah replied:
I added some comments, I hope they are enough

az108 replied:
yes thanks

az108 commented on src/main/webapp/app/pages/application/application-overview-for-applicant/application-overview-for-applicant.component.ts:
```diff
@@ -0,0 +1,109 @@
+
+  constructor() {
+    effect(() => {
+      this.applicationService.getApplicationPagesLength().subscribe(val => this.total.set(val));
```
we should have a look if its possible without subscribe. I can also look later

Douglas-Sarah replied:
I changed it like this:
```
  constructor() {
    effect(() => {
      this.loadTotal();
    });
  }

  async loadTotal(): Promise<void> {
    try {
      const tempTotal = await firstValueFrom(this.applicationService.getApplicationPagesLength());
      this.total.set(tempTotal);
    } catch (error) {
      alert('Failed to load total application count' + error);
    }
  }
  ```
  Do you think that looks better?

az108 commented on src/main/java/de/tum/cit/aet/application/domain/dto/ApplicationOverviewDTO.java:
```diff
@@ -0,0 +1,67 @@
+     * @param createdAt The creation timestamp of the job.
+     * @return A string like "Today", "3 days ago", "2.5 years ago".
+     */
+    public static String getRelativeTime(Instant createdAt) {
```
please move this to a more generalized section/folder as this is the same as ishanis getRelativeTime. If we dont have one yet we should maybe make one

Douglas-Sarah replied:
I moved it into a shared service in `core` 👍

az108 reviewed (changes requested):
some comments

az108 reviewed (approved):
LGTM
